Environment variables

federico-paolillo opened the pull request:
This PR introduces support to environment variables as fallback values for command options.  

The interface `IEnvironmentVariablesProvider` has been introduced to abstract away environment variables access. Given the very general interface methods, this interface has the side effect of allowing variables to come from anywhere (e.g.: a file, Redis, a database), enabling flexibility in what can be considered an environment by an implementor.  

I decided to avoid putting all system variables on `CommandInput` and instead I decided to acces environment variables only on demand and when the name is known.    
Using `CommandInput` would have required to put **all** the env. variables on and instance of `CommandInput` because the actual environment variable names required are known only on the `CommandSchema` and every `CommandOptionSchema`. Additionally exposing all the environment variables of a system can be a security risk as they may contain secrets that could be leaked in logging. 

To access environment variables by name a `CommandOptionSchema` instance is needed in oreder to read the property `EnvironmentVariableName` which, as the name implies, contains the needed variable name.  
I decided to extend `CommandInitializer` by "inverting" the foreach loop: insteand of going through every `CommandOptionInput` the loop now goes through every `CommandOptionSchema` and in case a given schema does not have a `CommandOptionInput` environment variables are queried for values. Once values are found they are turned into a substitutive `CommandOptionInput` and processed as always.  

To allow flexibility and guarantee testability I've also extended `IApplicationBuilder` to let an user specify a custom `IEnvironmentVariablesProvider`.

There a few open points before considering this PR complete, in particular:
- What to document on README about this feature
- How help text should look like for options with env. variables support
- `SecurityException`s raised when accessing variables are ignored for now

Closes #3

federico-paolillo commented:
I believe CodeCov complains about missing coverage in `EnvironmentVariablesProvider`.  
I did not test that class because it would be a very hefty integration test that would require setting environment variables in the test system.

I'm not really sure how to proceed, is it possible to ignore code coverage for specific files ?

Tyrrrz commented on CliFx.Tests/CliApplicationBuilderTests.cs:
```diff
@@ -1,8 +1,9 @@
+using CliFx.Tests.Stubs;
 using CliFx.Tests.TestCommands;
 using NUnit.Framework;
+using System;
```
Please keep the previous namespace ordering: System. first (here and other places)

federico-paolillo replied:
Visual Studio format document 😑 I'll fix it

Tyrrrz commented on CliFx/Services/CommandInitializer.cs:
```diff
@@ -38,15 +48,25 @@ public void InitializeCommand(ICommand command, CommandSchema commandSchema, Com
-                if (optionSchema == null)
-                    continue;
+                //Find matching option input
+                CommandOptionInput optionInput = commandInput.Options.FindByOptionSchema(optionSchema);
```
Use var please

Tyrrrz commented on CliFx/Services/EnvironmentVariablesProvider.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <summary>
+        /// Default delimiter for multiple values
+        /// </summary>
+        public const char DEFAULT_VALUES_DELIMITER = ',';
```
Looks like Windows is using ';' as delimiter (e.g. `$env:PATH`)

Tyrrrz replied:
Probably should also account that ';' might be a valid character inside an environment variable... Is there some kind of convention for escaping values here?

federico-paolillo replied:
We can use [Path.PathSeparator](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.pathseparator?view=netstandard-2.0), looks like the env. specific separator for multiple values in an env. variable.
[According to this](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/set_1) the escape character is `^`

Tyrrrz replied:
Ha. Interesting. Yeah, let's use that. Let's not worry about escapes for now,
Regarding escapes, it seems that, similarly to command line arguments, environment variable enclosed in quotes is treated as one value. Maybe we should handle that?

E.g. `"value1;value2" is still treated as one value while `value1;value2` isn't. We can leave this for later.

Tyrrrz commented on CliFx.Tests/Services/CommandInitializerTests.cs:
```diff
@@ -69,7 +70,34 @@ private static IEnumerable<TestCaseData> GetTestCases_InitializeCommand()
+                new EnvironmentVariableCommand(),
+                GetCommandSchema(typeof(EnvironmentVariableCommand)),
+                new CommandInput(null, new CommandOptionInput[] { }),
+                new EnvironmentVariableCommand { Option = "A" }
```
It would be nice to have a test that ensures that, when property type is not a collection, environment variable is always treated as single value.

Example:

Option of type `string`.
Environment variable: `value1,value2`
Resulting option value: `"value1,value2"`

Option of type `string[]`
Environment variable: `value1,value2`
Resulting option values: `[] {"value1", "value2"}`

Does this make sense?

federico-paolillo replied:
Got it, makes sense

Tyrrrz commented on CliFx/Services/EnvironmentVariablesProvider.cs:
```diff
@@ -0,0 +1,57 @@
+        public IReadOnlyList<string> GetValues(string variableName)
+        {
+            //If variableName is null simply return nothing, this may happen if a Command has not EnvironmentVariable fallback
+            if (variableName == null) return null;
```
If option does not have environment variable fallback, `CommandInitializer` shouldn't be calling this method

federico-paolillo replied:
Also true, I'll make sure of that

Tyrrrz commented on CliFx/Services/CommandInitializer.cs:
```diff
@@ -38,15 +48,25 @@ public void InitializeCommand(ICommand command, CommandSchema commandSchema, Com
+                CommandOptionInput optionInput = commandInput.Options.FindByOptionSchema(optionSchema);
+
+                //If no option input is available fall back to environment variable values
+                if (optionInput == null)
```
Also check that `!optionSchema.EnvironmentVariableName.IsNullOrWhiteSpace()`

Tyrrrz commented on CliFx/Services/EnvironmentVariablesProvider.cs:
```diff
@@ -0,0 +1,57 @@
+            {
+                string variableValue = Environment.GetEnvironmentVariable(variableName);
+
+                if (string.IsNullOrWhiteSpace(variableValue)) return null;
```
Can use extension method here

federico-paolillo replied:
You mean something like `Environment.GetEnvironmentVariableOrDefault()` ?

Tyrrrz replied:
No, I mean there's extension `variableValue.IsNullOrWhiteSpace` which is a bit shorter :)

Tyrrrz commented on CliFx/Services/EnvironmentVariablesProvider.cs:
```diff
@@ -0,0 +1,57 @@
+
+            try
+            {
+                string variableValue = Environment.GetEnvironmentVariable(variableName);
```
Use var please

Tyrrrz commented on CliFx/Services/EnvironmentVariablesProvider.cs:
```diff
@@ -0,0 +1,57 @@
+
+                return values.ToList();
+            }
+            catch (SecurityException)
```
When does this happen, out of curiosity?

federico-paolillo replied:
According to Microsoft documentation it says that 

> The caller does not have the required permission to perform this operation.

I believe this might happen in Windows if an user is under some group policy that forbids reading some env. variables although that's speculation but I tought that given that the documentation reports it it is always better to handle it even if by simply swallowing it and returning `null`.

Tyrrrz replied:
Hm. It will be hard to cover this. Maybe throwing an exception makes sense, the user would want to know why environment variable is not resolved. It will only be called if `EnvironmentVariableName` is set to it shouldn't interfere with other options.

Tyrrrz commented on CliFx/Services/CommandSchemaResolver.cs:
```diff
@@ -31,7 +31,8 @@ private IReadOnlyList<CommandOptionSchema> GetCommandOptionSchemas(Type commandT
                     attribute.IsRequired,
-                    attribute.Description);
+                    attribute.Description,
+                    attribute.EnvironmentVariableName);
```
Will also need a check to make sure there are no options with duplicate environment variable names?

federico-paolillo replied:
I cannot image an use case where binding multiple options to the same environment variable could be usefule. I guess it makes sense to forbid to use the same variable multiple times, although, given that it does not affect options parsing, we could leave it as it is for additional flexibility.

Tyrrrz replied:
Yeah I can imagine it could be used purposely, although it's more likely a user error. Let's not validate it for now.

Tyrrrz commented:
Don't worry regarding coverage. It looks like you dropped 7% in `CommandInitializer` but there are only 3 uncovered statements (constructor overload).

federico-paolillo commented:
I'll get to work in the next days to fix all the requested changes and will report back.  

I cannot really argue that env. are more than enough for a console application so the additional flexibility might be wasted (apart for testing). Should i try to rewrite the logic to use `CommandInput` instead of  `IEnvironmentVariablesProvider` and see how it looks ?

Tyrrrz commented:
>I'll get to work in the next days to fix all the requested changes and will report back.

Awesome, thanks :)

>I cannot really argue that env. are more than enough for a console application so the additional flexibility might be wasted (apart for testing). Should i try to rewrite the logic to use CommandInput instead of IEnvironmentVariablesProvider and see how it looks?

If you have time, that would be nice. If not, I'll just tackle this later.

federico-paolillo commented:
Finally got time to fix everything.

As dicussed environment variables are now provided via `CommandInput.EnvironmentVariables`, to get variables `CommandInputParser` still calls `IEnvironmentVariablesProvider`. This is needed to make the implementation testable, calling `Environment.GetEnvironmentVariables` directly would make impossible to test `CommandInputParser`. 

I've introduced a new class called `EnvironmentVariablesParser` that is responsible to process an environment variable value and parse it so that it matches the criteria discusses before (e.g.: split values only when an option is a collection, etc...). This class is invoked by `CommandInitializer` because in order to parse variables correctly we must know for what `OptionSchema` we must provide values.  

I've decided that in case a `SecurityException` is raised it is simply swallowed, rethrowing the exception wrappend in a `CliFxException` is not suitable because it would force applications that do not use environment variables to have access to them because `CommandInputParser` does not know in advance if an option requires variables or not and will access variables anyway.

`EmptyEnviromentVariablesStub` has been introduced to avoid creating many constructor overloads for `CommandInput` and to keep testing code as close to before as possible.

Escaping variables is not supported at the moment.

Tyrrrz commented:
Thank you for your effort!
I'll merge it now and make a few cosmetic changes later so you don't have to spend your time on this. :)
